ciphey/iface/_modules.py

```diff
@@ -167,6 +167,21 @@ def __init__(self, config: Config):
         return PolyWrapperClass
 
 
+class Tagger(Generic[T], ConfigurableModule, Targeted):
```

Comment:
@Cyclic3 here too

Reply:
Is this class okay?

Reply:
What does it do?

Reply:
Nothing at the moment, it's just I wanted to lay the groundwork.

The idea is you feed it the user input "Encoding and Popular not numberBased" and it'll spit out a set of modules which match that expression.  And then in Searcher, I was going to do something easy like:

```
If module not in tagged_modules:
    skip this module
```

Reply:
Surely there can only be one implementation of this? All the work would be in the translation of the conditions. Maybe a list of tuples (tag, bool) being passed to a function would be smoother.

Reply:
We'd need to extract the code for the tagging itself somehow, as it's the Clean Code:tm: way to extract logic. I saw the resource loader in basemods and assumed this was the best place to put it :sweat_smile: 

Where would you suggest placing a tagging system in Ciphey?

Reply:
My suggestion would be to add a static `getTags() -> Set[str]` method to configurable module, add a `tags` parameter to `Searcher.search` (defaulting to none, with a white/blacklist) , and add cli flags that map to it.

When `tags is not None`, the searcher checks each module to make sure it contains at least one element of the whitelisted, and no elements of the blacklist.

Reply:
Right, but to get that whitelist / blacklist we need a module to calculate that -- as we plan to support `union`, `intersection`, `or`, and any other set theory I can think of. 

Is this the right location for that specific module?

It's cleaner coding for the searcher to do one thing (search) rather than 2 things (search and generating whitelist / blacklist from tags)

Reply:
Intersection ≈ and, union ≈ or. In fact, the propositional calculus symbols for and/or are basically more pointy versions of the set theory ones for union/intersecion:

|symbol|definition|
| -| -|
|∧| OR |
| ⋂ | intersection |
|∨ | AND |
| ⋃ | union |

If you want, I can demonstrate this more formally than "similar symbols == the same", but this is definitely the case.

`</maths>`

A user is only really likely to want to exclude some modules, and include other ones. A whitelist/blacklist feature allows us to handle this. 

In my view, the searcher's job is already to select which decoders/crackers get called, and so I do not see this as an departure from its existing features.

Reply:
We would have to copy and paste the code for the tag system across searchers, which defines it as a separate problem. Different searchers can have different tag systems, when really there is only one tag system.

To me, the searchers job is to do what it says on the tin -- "search". Not generate the nodes needing to be searched. I can get what you're coming at, but I feel like it'd be better for us to at least make it a separate file in searchers to ensure all searchers all use the same tag system without needing to copy and paste each one.

Reply:
@bee-san The entire role of "searching" is generating the nodes that need to be searched, as that is the thing that takes time. What we want is a way of telling the searcher to not bother using some nodes. 

If we do this at the registry level, we kill our ability to be a library, as we would have global mutable state, meaning that if a program used two libraries that depended on ciphey, it would break.

As I see it, we have two choices:

1. Add an argument to the various lookup methods in the registry that filters out the result, which would *have* to be passed to the Searcher (or else it wouldn't know what to do), perhaps by means of the config object. 
1. Add an argument to the searcher directly, and let it do the filtering, again by means of the config object.